Timeout & LowPowerTimeout tests

fkjagodzinski opened the pull request:
## Description
Added unit tests for timeout classes
* `Timeout`,
* `LowPowerTimeout`.

## Status
**READY**

Preceding PR's needed:
* ~#5046~ -- provides a fix for several boards failing to handle a delay of 0 us,
* ~#5347~ -- provides a fix for #5150,

Issues to be resolved before merge:
* ~#5159~
* ~#5431~ 
* ~#5284~

0xc0170 commented on TESTS/mbed_drivers/timeout/main.cpp:
```diff
@@ -1,44 +1,196 @@
+    sem->release();
+}
+
+void timeout_inc_callback(volatile uint32_t *cnt) {
```
coding style alignment in this file

fkjagodzinski replied:
Ok, I fixed brace positions for function declarations.

0xc0170 commented:
cc @pan- @c1728p9

theotherjimmy commented:
/morph test

theotherjimmy commented:
/morph test-nightly

fkjagodzinski commented:
As suggested in [#5163](https://github.com/ARMmbed/mbed-os/pull/5163#pullrequestreview-65534381) I've made an update to avoid code duplication:
* rebased onto master (c60194fdf),
* extracted test case templates to a header file,
* added unit tests for `LowPowerTimeout`.

fkjagodzinski commented on TESTS/mbed_drivers/lp_timeout/main.cpp:
```diff
@@ -119,29 +121,48 @@ void lp_timeout_500us(void)
+    Case("Callback override (with attach_us)", test_override<AttachUSTester<LowPowerTimeout> >),
+
+    Case("Multiple timeouts running in parallel (with attach)", test_multiple<AttachTester<LowPowerTimeout> >),
+    Case("Multiple timeouts running in parallel (with attach_us)", test_multiple<AttachUSTester<LowPowerTimeout> >),
```
These two tests with multiple `LowPowerTimeout` objects fail on _K64F_, _NUCLEO_F070RB_, _NUCLEO_L476RG_. Waiting for the outcome of #5150.

fkjagodzinski commented on TESTS/mbed_drivers/lp_timeout/main.cpp:
```diff
@@ -119,29 +121,48 @@ void lp_timeout_500us(void)
+    Case("Multiple timeouts running in parallel (with attach_us)", test_multiple<AttachUSTester<LowPowerTimeout> >),
+
+    Case("Zero delay (with attach)", test_no_wait<AttachTester<LowPowerTimeout> >),
+    Case("Zero delay (with attach_us)", test_no_wait<AttachUSTester<LowPowerTimeout> >),
```
These tests fail on _NUCLEO_F070RB_, _NUCLEO_F429ZI_ and _NUCLEO_L476RG_ -- issue #5159.

0xc0170 commented:
Thanks for the report, I'll label this as needs preceeding PR, and we need to wait for the fixes until this is made green

bulislaw commented on TESTS/mbed_drivers/lp_timeout/main.cpp:
```diff
@@ -119,29 +121,48 @@ void lp_timeout_500us(void)
+    Case("Zero delay (with attach)", test_no_wait<AttachTester<LowPowerTimeout> >),
+    Case("Zero delay (with attach_us)", test_no_wait<AttachUSTester<LowPowerTimeout> >),
+
     Case("500us LowPowerTimeout", lp_timeout_500us, greentea_failure_handler),
```
Can we make these tests common for both test suites?

fkjagodzinski replied:
Sure, I'll move these test cases to the header file.

fkjagodzinski commented:
> What do you think about moving the tests to
TESTS/mbed_drivers/timeout/{lp_timeout,timeout}/

@bulislaw That was my initial idea but as it turned out this would require changing test discovery procedure which now expects strictly defined dir structure https://github.com/ARMmbed/mbed-os/blob/373e6ab4fb6d950d2822bb24d66b5eea0e362c4b/tools/test_api.py#L2046-L2050 so I chose a simpler solution.

bulislaw commented:
> @bulislaw That was my initial idea but as it turned out this would require changing test discovery procedure which now expects strictly defined dir structure

ok, makes sense.

adbridge commented:
@fkjagodzinski @bulislaw is this now ready ?

fkjagodzinski commented:
> is this now ready ?

@adbridge PR needs a minor update. Should be ready tomorrow.

fkjagodzinski commented:
Update:
* rebased onto current master (0e1cc5691),
* moved **ALL** test case templates to a header file common for `Timeout` and `LowPowerTimeout`,
* squashed commits into a single one for readability,
* increased a tolerance of measured delay `SHORT_DELTA_US` to `1200 us` -- needed for _NRF51_DK_ board to pass these tests; considering new tolerance value I've also:
  * removed `"500 us delay accuracy"` test case,
  * removed `"1 ms delay accuracy"` test case,
  * added `"10 ms delay accuracy"` test case,
* updated `test_sleep` and `test_deepsleep` to re-enter `sleep()` in case of other interrupt waking up the uC.

0xc0170 commented:
/morph build

bulislaw commented:
/morph build

fkjagodzinski commented:
With #5046 merged most errors (12/14) from the latest build will be fixed as they are all the consequence of #5284 issue (handling a delay of 0 us).
Regarding 2 other errors I've created an issue for **NCS36510** -- #5431.

fkjagodzinski commented:
Update:
* rebased onto master, which is now at 8b6a7aa,
* further increased a tolerance of measured delay `SHORT_DELTA_US` to `2000 us` -- I have seen an error of `1597 us` on NRF51_DK during my local tests.

Now all the boards that failed before, that is NCS36510, NRF51_DK and NRF52_DK, pass every test case!

@bulislaw, @0xc0170 could you take alook and accept this PR again, please? No changes (apart that `SHORT_DELTA_US`) were introduced.

fkjagodzinski commented:
@bulislaw, @0xc0170, @cmonr Could you run morph tests please? I'm curious to see the results; there is still one issue left for NCS36510 (#5431) but this does not show up in my local tests using GCC_ARM toolchain and I don't have access to other toolchains at the moment.

bulislaw commented:
@fkjagodzinski we are very close to a code freeze for 5.8, only PR critical for 5.8 can be run on the CI till we freeze (hopefully tomorrow). Next week should look better, we'll run it then.

fkjagodzinski commented:
@0xc0170, @cmonr Could you update the labels with `needs: review` and ` needs: CI`, and/or run morph tests please?

cmonr commented:
@0xc0170 @bulislaw Could y'all take another look? It looks like this is ready to go again.

@0xc0170 Does this still need the DNM label?

0xc0170 commented:
/morph build

0xc0170 commented:
Restarting, a device error

/morph test
